### ISSUE TITLE
[Clang] Update `-Wstrict-prototypes`

### ISSUE THREAD

**Issue by Sirraide** [llvm/llvm-project#84497]
I noticed this a few days ago but forgot about it, so this is to remind me to fix this later. This code:
```c
int main() {}
```
compiled as C with `-std=c17` and `-pedantic` prints:
```c
test.c:4:9: warning: a function declaration without a prototype is deprecated in all versions of C [-Wstrict-prototypes]
    4 | int main() {
      |         ^
      |          void
```
‘deprecated in all versions of C’ is not exactly correct anymore; this should instead warn that this is a C23 extension.

**Comment by Sirraide** [comment]
Actually, this warning isn’t really *wrong* from a technical point of view, because that is a function with no prototype before C23, but it’s perhaps not that obvious what that actually means, so I still think the wording here should be changed.

**Comment by Sirraide** [comment]
Discussed this w/ Aaron, and the reason we include ‘is deprecated in all versions of C’ is because, apparently, a lot of people are still writing `()` for `(void)` w/o realising that that is in fact not a prototype in older versions of C, so that part of the warning really is there for a reason. We *could* maybe find a way to update it, but also, it’s not really a priority and we don’t want to dump an entire paragraph into a warning, so I think the warning is ultimately fine as-is.

**Comment by jmarshall** [comment]
I recently encountered this diagnostic in similar code, and was surprised because this is a function **definition** rather than just a declaration (as per the message). It's not obvious why there's any problem here for a function _definition_, in which there's surely no confusion about whether the function is meant to have parameters or not — can you point to some reading about what's changed in C23 to make this warning desirable for function definitions?

**Comment by Sirraide** [comment]
there's surely no confusion about whether the function is meant to have parameters or not

One would hope so, but also, this is C, so I wouldn’t be so sure about that; @AaronBallman probably knows more about that; he’s a lot more familiar w/ the intricacies of C than me (I mostly tend to stick to C++).

> what's changed in C23 to make this warning desirable for function definitions?

What’s changed is that, in C23, you can finally write `int foo()` to indicate that a function takes no parameters. I.e. this is no longer a prototypeless function as of C23. We could change the warning to indicate that, but also, Aaron has pointed out to me that a lot of people unknowingly still use `()` instead of `(void)` in versions of C before C23, which is why the warning says ‘is deprecated in all versions of C’ (including even C89, iirc).

Our consensus here was that, while we could change the warning, it probably isn’t worth it to print out an entire paragraph explaining why you shouldn’t be writing `()` before C23...

**Comment by jmarshall** [comment]
Thanks for the information.

I certainly agree about alerting people writing `()` in declarations that it actually means arguments-could-be-anything. However I found text in §6.7.6.3/14 supporting the idea that **definitions** have always been less vague — so I've raised a new issue (#90596) instead of further derailing this older already closed one.
